rbd/mirror_group.go

```diff
@@ -0,0 +1,493 @@
+	cGroupName := C.CString(groupName)
+	defer C.free(unsafe.Pointer(cGroupName))
+
+	rbdMirrorGroupResync, rbdMirrorGroupResyncErr := dlsym.LookupSymbol("rbd_mirror_group_resync")
+	if rbdMirrorGroupResyncErr != nil {
+		return fmt.Errorf("%w: %w", ErrNotImplemented, rbdMirrorGroupResyncErr)
+	}
+
+	ret := C.rbd_mirror_group_resync_dlsym(
+		rbdMirrorGroupResync,
+		cephIoctx(groupIoctx),
+		cGroupName)
+	return getError(ret)
+}
+
+// MirrorGroupState represents the current state of the mirrored group
+type MirrorGroupState C.rbd_mirror_group_state_t
+
+// String representation of MirrorGroupState.
+func (mgs MirrorGroupState) String() string {
+	switch mgs {
+	case MirrorGroupEnabled:
+		return "enabled"
+	case MirrorGroupDisabled:
+		return "disabled"
+	case MirrorGroupEnabling:
+		return "enabling"
+	case MirrorGroupDisabling:
+		return "disabling"
+	default:
+		return "<unknown>"
+	}
+}
+
+const (
+	// MirrorGroupDisabling is the representation of
+	// RBD_MIRROR_GROUP_DISABLING from librbd.
+	MirrorGroupDisabling = MirrorGroupState(C.RBD_MIRROR_GROUP_DISABLING)
+	// MirrorGroupEnabling is the representation of
+	// RBD_MIRROR_GROUP_ENABLING from librbd
+	MirrorGroupEnabling = MirrorGroupState(C.RBD_MIRROR_GROUP_ENABLING)
+	// MirrorGroupEnabled is the representation of
+	// RBD_MIRROR_IMAGE_ENABLED from librbd.
+	MirrorGroupEnabled = MirrorGroupState(C.RBD_MIRROR_GROUP_ENABLED)
+	// MirrorGroupDisabled is the representation of
+	// RBD_MIRROR_GROUP_DISABLED from librbd.
+	MirrorGroupDisabled = MirrorGroupState(C.RBD_MIRROR_GROUP_DISABLED)
+)
+
+// MirrorGroupInfo represents the mirroring status information of group.
+type MirrorGroupInfo struct {
+	GlobalID        string
+	State           MirrorGroupState
+	MirrorImageMode ImageMirrorMode
+	Primary         bool
+}
+
+// GetMirrorGroupInfo returns the mirroring info of the group
+//
+// Implements:
+//
+//	int rbd_mirror_group_get_info(rados_ioctx_t p, const char *name,
+//								  rbd_mirror_group_info_t *mirror_group_info,
+//								  size_t info_size)
+func GetMirrorGroupInfo(groupIoctx *rados.IOContext, groupName string) (*MirrorGroupInfo, error) {
+	var cgInfo C.rbd_mirror_group_info_t
+	cGroupName := C.CString(groupName)
+	defer C.free(unsafe.Pointer(cGroupName))
+
+	rbdMirrorGroupGetInfo, rbdMirrorGroupGetInfoErr := dlsym.LookupSymbol("rbd_mirror_group_get_info")
+	if rbdMirrorGroupGetInfoErr != nil {
+		return nil, fmt.Errorf("%w: %w", ErrNotImplemented, rbdMirrorGroupGetInfoErr)
+	}
+
+	ret := C.rbd_mirror_group_get_info_dlsym(
+		rbdMirrorGroupGetInfo,
+		cephIoctx(groupIoctx),
+		cGroupName,
+		&cgInfo,
+		C.sizeof_rbd_mirror_group_info_t)
+
+	var err error
+	if ret < 0 {
+		err = getError(ret)
+	}
+
+	if err != nil {
+		if errors.Is(err, ErrNotFound) || errors.Is(err, rados.RadosErrorNotFound) {
+			return &MirrorGroupInfo{
+				State:           MirrorGroupDisabled,
+				MirrorImageMode: ImageMirrorModeSnapshot,
+				Primary:         false,
+			}, nil
+		}
+	}
```

Comment:
I don't this should be handled in go-ceph ?

Reply:
Yes, this was a hack, it should be removed from go-ceph

Reply:
Yeah, removed the hack.